### PR TITLE
Introducing Threads in Java Socket Programming

### LINKED ISSUE #4554: [Networking] Introducing Threads in Java Socket Programming
#### Proposed title of article
Introducing Threads in Java Socket Programming
#### Proposed article introduction
We don't want just one client to connect to the server at a time, but many clients to connect at the same time. We want our architecture to be able to support multiple clients at once. As a result, threads must be used on the server side so that whenever a client request comes in, a separate thread can be assigned to handle each request.
Assume that two requests arrive at the server. With our basic server-client program, the request that arrives even a nanosecond first will be able to connect to the server, while the other request will be rejected because no mechanism for handling multiple requests concurrently is provided. Threading is used in network programming to solve this problem.
#### Key takeaways 
- Why use threads in-network programming
- Server-Side Programming
- Client-Side Programming

#### References
N/A
### Templates to use as guides
- [How To Guide Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/how-to-guide-template.md)
- [Software Review Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/software-review-template.md)
- [Tutorial Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/tutorial-template.md)

### PR DESCRIPTION
**Pull request by SteveadeNick**
### Article checklist

> If you have not already, please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve your overall technical writing so reviews are swifter, increase the chance of a payout and provide more value to our readers.

> For first-time contributors and for more details on our submission guidelines, **see our [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**.

#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [ ] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

This fixes #4554

### CODE REVIEW

**Comment by SteveadeNick** [comment]
This fixes #4554

**Comment by ahmadmardeni1** [comment]
Upon running your article through our 3rd party plagiarism checker it seemed to raise a few flags and the % was higher than we typically accept.
Please see attached PDF - and revisit the article to ensure we are contributing wholly unique and original content.
Be sure to see our resources page to see more info on plagiarism and what is considered as such. 
Also, Why did you delete our template before submitting your PR? @SteveadeNick 

[Introducing Threads in Java Socket Programming#4937](https://github.com/section-engineering-education/engineering-education/files/7547300/New.Scan.2_30.PM.Copyleaks.Report.pdf)

**Comment by SteveadeNick** [comment]
@ahmadmardeni1 
Hello
I have the revisions with me. Could you please reopen so as I can commit them for a final plagiarism check.
Thank you.

**Comment by ahmadmardeni1** [comment]
Just added the template. Please check what matched your aricle. @SteveadeNick

**Comment by SteveadeNick** [comment]
Thanks @ahmadmardeni1

**Comment by paulodhiambo** [comment]
@SteveadeNick please explain your code, at least add inline comments to explain parts of your code.

**Comment by SteveadeNick** [comment]
> @SteveadeNick please explain your code, at least add inline comments to explain parts of your code.

Done @paulodhiambo

**Comment by SteveadeNick** [comment]
Any progress with this article?

**Comment by SteveadeNick** [comment]
@paulodhiambo @ahmadmardeni1 
I revised this article.

File: content/articles/introducing-threads-in-java-socket-programming/index.md [hunk 1]
```diff
@@ -0,0 +1,442 @@
...
+
+- It is accepted by the server in the while loop after it has been assigned.
+
+- As long as the first request is still being processed, the server accepts and processes the second one. Even as other requests are being processed, many requests can be addressed simultaneously.
+### Testing the above program 
+It does thread matter where you save the two applications. Afterward, run the Server.java and the Client.java one after the other. In order to execute numerous instances of the client program at the same time, utilize an IDE like multiple instances, or run the client software in separate files. To get the results data input stream played above, you'll need numerous clients.
+### Conclusion
+In this article, we’ve learned socket programming, thread and its use, server-side programming, and also client-side programming.
+
+Our implementation showed the importance of threads in java and enhancing understanding of threads in java socket programming.
```
Comment (paulodhiambo) [inline content/articles/introducing-threads-in-java-socket-programming/index.md:440]: reword to "This article has learned socket programming, thread and its use, server-side programming, and client-side programming.
Our implementation showed the importance of threads in Java and enhancing understanding of threads in java socket programming."

File: content/articles/introducing-threads-in-java-socket-programming/index.md [hunk 2]
```diff
@@ -0,0 +1,442 @@
...
+### How these programs operate together
+- A new thread is started to connect to the server for each request made by a client. To communicate with the client, the newly assigned thread is given access to the available streams.
+
+- It is accepted by the server in the while loop after it has been assigned.
+
+- As long as the first request is still being processed, the server accepts and processes the second one. Even as other requests are being processed, many requests can be addressed simultaneously.
+### Testing the above program 
+It does thread matter where you save the two applications. Afterward, run the Server.java and the Client.java one after the other. In order to execute numerous instances of the client program at the same time, utilize an IDE like multiple instances, or run the client software in separate files. To get the results data input stream played above, you'll need numerous clients.
```
Comment (paulodhiambo) [inline content/articles/introducing-threads-in-java-socket-programming/index.md:436]: reword to "It does thread matter where you save the two applications. Afterward, run the Server.java and the Client.java one after the other. To execute numerous instances of the client program simultaneously, utilize an IDE like multiple instances, or run the client software in separate files. To get the results data input stream played above, you will need numerous clients."

File: content/articles/introducing-threads-in-java-socket-programming/index.md [hunk 3]
```diff
@@ -0,0 +1,442 @@
...
+    }
+}
+
+ ```
+ ### Program's method of operation
+To run the Java Server Socket Program, you must first launch it from the data output stream prompt (console). Then, you will see a message that reads "Server Started..." on your data output stream screen.
+
+Java Client Socket Programs must be opened on both computers or on the same network. Your client program will wait for your input when you start it up. If you input a message and press ENTER, the server will also data input stream play the message. You can send messages to the client from the server after receiving them from the client. When the client transmits the word "bye" from the client-side of the connection, the connection is terminated.
+### How these programs operate together
+- A new thread is started to connect to the server for each request made by a client. To communicate with the client, the newly assigned thread is given access to the available streams.
```
Comment (paulodhiambo) [inline content/articles/introducing-threads-in-java-socket-programming/index.md:430]: reword to " A new thread is started to connect to the server for each request made by a client. For a connection to be established, the newly assigned thread is given access to the available streams."

File: content/articles/introducing-threads-in-java-socket-programming/index.md [hunk 4]
```diff
@@ -0,0 +1,442 @@
...
+### Why use of threads
+It's also important that the server can handle multiple clients at once.Having the ability to handle several consumers simultaneously is a requirement of our design framework. Our server side implementation needs threading so that each request may be assigned a data input stream thread as it comes in.
+There are two methods for creating a new thread in Java:
+
+- Implementation of the Runnable interface is required. It is possible to run (java.lang.Runnable)
+- There are many ways to extend the Thread class. (java.lang.Thread)
+### Server-Side Programming
+Server Socket Program is a Java Console Application.
+**Server class:** Client-side construction begins with a collection of streams and port numbers from the server, and then follows these steps:
+- A while loop is used to build socket objects that receive connections from server socket objects.
```
Comment (paulodhiambo) [inline content/articles/introducing-threads-in-java-socket-programming/index.md:237]: reword to "**Server class:** Client-side construction begins with a collection of streams and port numbers from the server and then follows these steps:
- A while loop is used to build socket objects that receive connections from server socket objects."

File: content/articles/introducing-threads-in-java-socket-programming/index.md [hunk 5]
```diff
@@ -0,0 +1,442 @@
...
+
+Our implementation showed the importance of threads in java and enhancing understanding of threads in java socket programming.
+
+Happy learning!
+
+### Threads
+Each sequential program begins, progresses through its execution, and ends. In a thread, the program's control travels in a logical order. It's a separate path through the program's source code. When a high number of events or actions need to occur at the same time, many applications are developed to function on a single thread. It is common for one thread to take a different path through the same code as the others when numerous threads are running at the same time.Java.lang is responsible for creating and managing every thread in Java.Class of threads.
+### Why use of threads
+It's also important that the server can handle multiple clients at once.Having the ability to handle several consumers simultaneously is a requirement of our design framework. Our server side implementation needs threading so that each request may be assigned a data input stream thread as it comes in.
+There are two methods for creating a new thread in Java:
```
Comment (paulodhiambo) [inline content/articles/introducing-threads-in-java-socket-programming/index.md:230]: reword to "It's also essential that the server can handle multiple clients at once. Having the ability to manage several consumers simultaneously is a requirement of our design framework. Our server-side implementation needs threading to assign each request a data input stream thread as it comes in."

File: content/articles/introducing-threads-in-java-socket-programming/index.md [hunk 6]
```diff
@@ -0,0 +1,442 @@
...
+### Conclusion
+In this article, we’ve learned socket programming, thread and its use, server-side programming, and also client-side programming.
+
+Our implementation showed the importance of threads in java and enhancing understanding of threads in java socket programming.
+
+Happy learning!
+
+### Threads
+Each sequential program begins, progresses through its execution, and ends. In a thread, the program's control travels in a logical order. It's a separate path through the program's source code. When a high number of events or actions need to occur at the same time, many applications are developed to function on a single thread. It is common for one thread to take a different path through the same code as the others when numerous threads are running at the same time.Java.lang is responsible for creating and managing every thread in Java.Class of threads.
```
Comment (paulodhiambo) [inline content/articles/introducing-threads-in-java-socket-programming/index.md:227]: reword to "Each sequential program begins, progresses through its execution, and ends. In a thread, the program's control travels in a logical order. It's a separate path through the program's source code. When a high number of events or actions need to coincide, many applications are developed to function on a single thread. It is common for one thread to take a different path through the same code as the others when numerous threads are running at the same time.Java.lang is responsible for creating and managing every thread in `Java.Class` of threads."

**Review by paulodhiambo — CHANGES_REQUESTED** [review]
Hello @SteveadeNick please make this changes, I'll close this article for now and reopen once you are done making the changes to clear the queue.

**Comment by SteveadeNick** [comment]
@paulodhiambo I was in the middle of editing the article before you even started.

**Comment by SteveadeNick** [comment]
Screenshot (34)

**Comment by SteveadeNick** [comment]
Screenshot (35)

Also, I was adding the table of contents and the prerequisites section.

**Comment by SteveadeNick** [comment]
@paulodhiambo You can reopen. Thanks in advance.

**Comment by SteveadeNick** [comment]
@paulodhiambo 
Hello.
I have the changes with me, just waiting to be committed. Please reopen so that I can push them and ensure I have revised the article once more so that there will be no more revisions. Thank you in advance.

**Comment by SteveadeNick** [comment]
@paulodhiambo should I reopen a new PR?

**Comment by SteveadeNick** [comment]
@paulodhiambo Done

**Comment by hectorkambow** [comment]
@paulodhiambo can we get a status on this PR? 

It would be best for the author to submit the polished article via a PR ifi any more edits are needed. 

Or if its ready for a final let us know 👍

**Comment by paulodhiambo** [comment]
@SteveadeNick Please go through your article, there are several sentences that are incomplete. Check on how I have edited the first section of your article. For any assistance, ping me on slack.

**Comment by SteveadeNick** [comment]
Can you please reopen @paulodhiambo

**Comment by SteveadeNick** [comment]
@paulodhiambo should I a new PR?

**Comment by paulodhiambo** [comment]
@SteveadeNick please open a new PR and ping me.

**Comment by SteveadeNick** [comment]
@hectorkambow closed the topic i suggested

**Comment by paulodhiambo** [comment]
No problem, I'll reopen an link the issue

**Pull request closed without merging**